src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java

```diff
@@ -701,13 +769,82 @@ private void emitUnboxAddress() {
     private void emitBufferLoad(Binding.BufferLoad bufferLoad) {
         Class<?> loadType = bufferLoad.type();
         long offset = bufferLoad.offset();
+        int byteWidth = bufferLoad.byteWidth();
 
         popType(MemorySegment.class);
 
-        Class<?> valueLayoutType = emitLoadLayoutConstant(loadType);
-        emitConst(offset);
-        String descriptor = methodType(loadType, valueLayoutType, long.class).descriptorString();
-        emitInvokeInterface(MemorySegment.class, "get", descriptor);
+        if (SharedUtils.isPowerOfTwo(byteWidth)) {
+            Class<?> valueLayoutType = emitLoadLayoutConstant(loadType);
+            emitConst(offset);
+            String descriptor = methodType(loadType, valueLayoutType, long.class).descriptorString();
+            emitInvokeInterface(MemorySegment.class, "get", descriptor);
+        } else {
+            // chunked
+            int readAddrIdx = newLocal(MemorySegment.class);
```

Comment:
It's nice that, when we specialize we can get rid of the loop we have when interpreting - and just emit a sequence of load/stores (with some extra shifts).